### PR TITLE
Refactoring Token

### LINKED ISSUE #280: Refactoring | make the tokens and balances more EVM agnostic
## Description
Currently, we have a very specific type of Token (EvmToken) to manage all attributes related to EVM token balances and not pay attention to the Native Token properties. Therefore, we are creating functions specific to EvmTokens instead of having a more agnostic solution.

We need to discuss what should be a better typing for Tokens (NativeToken & EvmToken) and Balances (maybe TokenBalance?).

## Acceptance Criteria
- define a better token & balance typing
- avoid using specific classes as much as it can be done
- refactor the code by applying the new typing 

## Note
The balance store has a lot to benefit from this new refactoring.

### PR DESCRIPTION
**Pull request by Viterbo**
# Fixes #280

## Description
This is just a proposal for better implementation of the Token concept, currently represented as a couple of interfaces with repeated concepts in different attributes like address-contract, amount-balance, decimals-precision or logo-logoURI

This PR implements a simple class hierarchy merging all source formats into one unified class, which can be enhanced with as many utility functions as we need in the process of developing this new platform.

This is a diagram of the new classes to represent tokens and balances.

TokenClass drawio (2)

## Checklist:

-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] I have cleaned up the code in the areas my change touches
-   [x] My changes generate no new warnings
-   [x] Any dependent changes have been merged and published in downstream modules
-   [x] I have checked my code and corrected any misspellings
-   [x] I have removed any unnecessary console messages
-   [x] I have included all english text to the translation file
-   [x] I have created a new issue with the required translations for the currently supported languages
-   [ ] I have added appropriate test coverage

### CODE REVIEW

**Comment by Viterbo** [comment]
This PR is actually blocked because it was developed on top of the last not-merged-yet @EJSG's PR.
It would be a good thing to wait until [that PR](https://github.com/telosnetwork/telos-wallet/pull/332) gets approved and merged.

**Review by EJSG — APPROVED** [review]
very nice, this is much better. and i love the graph haha

**Comment by Viterbo** [comment]
> Questions on the graph, where is the dimension of the price kept so what fiat is it, USD, ETH, PESO whatever? Shouldn't there be many prices for each token, one in each possible currency (not only fiat but every other currency out there eth, btc ...)?

What a nice question!!! because I can extract a hidden requirement for the following function:
```typescript
TokenPrice.getAmountInThisToken(amount, targetToken): BN
```
and with this new function, I can answer your question:

All Tokens have a unique price and that price can be expressed in any other currency/token.
In this implementation, the information about the token price is hidden behind the `TokenMarketData.info` so it can be anything (currently covers the indexer response). Once we have that info for the token, the `token.price.isAvailable` variable becomes true and then you would be able to ask the following:

```typescript
const otherToken = await this.getToken('ETH');
const converted = balance.token.price.getAmountInThisToken(
   balance.amount,
   otherToken,
);
// converted is a BigNumber containing the amount of ETH the balance is worth
```

I will add this function right now !!

File: src/antelope/chains/EVMChainSettings.ts [hunk 1]
```diff
@@ -190,91 +196,65 @@ export default abstract class EVMChainSettings implements ChainSettings {
...
-                fullBalance: '',
-                fiatBalance: '',
-                // actual token values
-                ...t,
-            })))
-            .then(tokens => tokens.map(t => t as unknown as EvmToken));
+            }) as unknown as TokenSourceInfo))
+            .then(tokens => tokens.map(t => new TokenClass(t)))
+
+            // Por último concatenamos el token del sistema this.getSystemToken()
```
[the remark this answers is not in the document]
Comment (Viterbo) [reply src/antelope/chains/EVMChainSettings.ts:256]: jajaja que vergüenza :see_no_evil:

File: src/antelope/chains/EVMChainSettings.ts [hunk 2]
```diff
@@ -190,91 +196,65 @@ export default abstract class EVMChainSettings implements ChainSettings {
...
+        if (full !== undefined) {
+            aux = { full, ...aux };
         }
 
         const params: AxiosRequestConfig = aux as AxiosRequestConfig;
+        const url = `v1/address/${address}/transactions`;
 
-        return this.hyperion.get('/v2/evm/get_transactions', { params })
-            .then(response => response.data.transactions as EvmTransaction[]);
+        return await this.indexer.get(url, { params })
```
[the remark this answers is not in the document]
Comment (Viterbo) [reply src/antelope/chains/EVMChainSettings.ts:236]: I will argue for the opposite. In fact, we should remove the `await` to make it even better.

By using `await`, you are making the entire application pause waiting for the resolution of a single promise. If we abuse the use of this statement, we could cause some long processes such as the initial load to take longer than necessary by preventing independent processes from being parallelized and necessarily having to be executed in sequence (with await).

Using **pipes** is the best solution for processes that require several waits at various points. You just specify what is to be done next (using .then(...)) and that execution will be executed at some point (the most appropriate) and in the midtime we handle the promise of its response, to which other actions can be chained (also using .then(...)).

For a better understanding of the code I would suggest using comments

File: src/antelope/chains/evm/telos-evm/index.ts [hunk 3]
```diff
@@ -95,24 +98,24 @@ export default class TelosEVMTestnet extends EVMChainSettings {
...
+    getWrappedNativeTokenAddress() {
         return '0xD102cE6A4dB07D247fcc28F366A623Df0938CA9E';
     }
 
     getImportantTokensIdList(): string[] {
         return [
             this.constructTokenId(TOKEN),
-            this.constructTokenId({ symbol: 'STLOS', address: this.getStlosContractAddress() } as EvmToken),
-            this.constructTokenId({ symbol: 'WTLOS', address: this.getWtlosContractAddress() } as EvmToken),
+            this.constructTokenId({ symbol: 'STLOS', address: this.getStakedNativeTokenAddress() } as TokenSourceInfo),
```
[the remark this answers is not in the document]
Comment (Viterbo) [reply src/antelope/chains/evm/telos-evm/index.ts:112]: Good catch !!

File: src/antelope/chains/EVMChainSettings.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmjanus) [reply src/antelope/chains/EVMChainSettings.ts]: I get it, code is easier to understand without the await as well.

**Comment by Viterbo** [comment]
I'm closing this PR in favor of this other new PR because it was easy that rebase with the current develop branch
https://github.com/telosnetwork/telos-wallet/pull/362

**Pull request closed without merging**